src/goLanguageServer.ts

```diff
@@ -0,0 +1,377 @@
+
+import vscode = require('vscode');
+import path = require('path');
+import {
+	LanguageClient, RevealOutputChannelOn, FormattingOptions, ProvideDocumentFormattingEditsSignature,
+	ProvideCompletionItemsSignature, ProvideRenameEditsSignature, ProvideDefinitionSignature, ProvideHoverSignature,
+	ProvideReferencesSignature, ProvideSignatureHelpSignature, ProvideDocumentSymbolsSignature, ProvideWorkspaceSymbolsSignature,
+	HandleDiagnosticsSignature, ProvideDocumentLinksSignature,
+} from 'vscode-languageclient';
+import { ProvideTypeDefinitionSignature } from 'vscode-languageclient/lib/typeDefinition';
+import { ProvideImplementationSignature } from 'vscode-languageclient/lib/implementation';
+import { GO_MODE } from './goMode';
+import { getToolFromToolPath } from './goPath';
+import { getToolsEnvVars } from './util'
+import { getCompletionsWithoutGoCode, GoCompletionItemProvider } from './goSuggest';
+import { GoHoverProvider } from './goExtraInfo';
+import { GoDefinitionProvider } from './goDeclaration';
+import { GoReferenceProvider } from './goReferences';
+import { GoImplementationProvider } from './goImplementations';
+import { GoTypeDefinitionProvider } from './goTypeDefinition';
+import { GoDocumentFormattingEditProvider } from './goFormat';
+import { GoRenameProvider } from './goRename';
+import { GoDocumentSymbolProvider } from './goOutline';
+import { GoSignatureHelpProvider } from './goSignature';
+import { GoWorkspaceSymbolProvider } from './goSymbol';
+import { parseLiveFile } from './goLiveErrors';
+import { promptForMissingTool } from './goInstallTools';
+import { getBinPath, getCurrentGoPath } from './util';
+
+interface LanguageServerConfig {
+	enabled: boolean;
+	flags: string[];
+	features: {
+		completion: boolean;
+		diagnostics: boolean;
+		format: boolean;
+		definition: boolean;
+		typeDefinition: boolean;
+		hover: boolean;
+		references: boolean;
+		rename: boolean;
+		signatureHelp: boolean;
+		documentSymbols: boolean;
+		workspaceSymbols: boolean;
+		implementation: boolean;
+		documentLink: boolean;
+	};
+}
+
+// registerLanguageFeatures registers providers for all the language features.
+// It looks to either the language server or the standard providers for these features.
+export function registerLanguageFeatures(ctx: vscode.ExtensionContext) {
+	// Subscribe to notifications for changes to the configuration of the language server.
+	ctx.subscriptions.push(vscode.workspace.onDidChangeConfiguration(e => watchLanguageServerConfiguration(e)));
+
+	let config = parseLanguageServerConfig();
+
+	// If the user has not enabled the language server,
+	if (!config.enabled) {
+		registerUsualProviders(ctx);
+		return;
+	}
+
+	let path = getLanguageServerToolPath();
+	const toolName = getToolFromToolPath(path);
+	const c = new LanguageClient(
+		toolName,
+		{
+			command: path,
+			args: ['-mode=stdio', ...config.flags],
+			options: {
+				env: getToolsEnvVars(),
+			},
+		},
+		{
+			initializationOptions: {},
```

Comment:
The reason to skip `funcSnippetEnabled` and `incrementalSync` from initialization options is?

Reply:
We don't actually respect those when passed into `gopls` so they have no effect right now. `incrementalSync` is already enabled by default, so the correct setting would actually be `noIncrementalSync`, but that's only useful for clients that don't support sending diffs. We are still working on a more thorough way to pass in configs, so I'll send a follow-up PR when we work out something that will be effective for all editors.

Reply:
makes sense, thanks!